src/core/net/dns_headers.cpp

```diff
@@ -293,7 +293,7 @@ otError Name::ReadName(const Message &aMessage,
     return error;
 }
 
-otError Name::LabelIterator::GetNextLabel(void)
+otError Name::LabelIterator::GetNextLabel(bool aStopOnEndOffsetSet)
```

Comment:
This `aStopOnEndOffsetSet` looks a little bit strange to me. 
I think the better solution might be always specifying header offset when creating a `LabelIterator`. 

Another thought is that I think **always making sure that header offset is at zero** can be a good practice for DNS/SRP server/client logic. For example, we won't need to specify `aHeaderOffset` for some DNS utility methods, which would reduce code size. It's actually quite simple to guarantee that the DNS header is at offset 0:
- When we receive a UDP DNS message, we remove the DNS header.
- When we create a DNS message to send, we make sure DNS header is at the beginning.

Thoughts? @abtink @wgtdkp

Reply:
My preference for now is to merge this fix to unblock the rest of the PRs. Changing this now with all pending and upcoming PRs would be rather messy.

For short-term (fix the behavior now):
- I don't favor having to provide `aHeaderOffset` to `ParseName()` (since we don't really need)
- Another approach: If you recall, in the initial version of `GetNextLabel()` we had a model to return `OT_ERROR_PENDING` when we got to "pointer label" to allow caller to decide to stop. We then removed this to optimize/simplify the code. We can use this instead of this PR
- I prefer the fix in this PR since it is simpler (and more clear).

For longer-term:
- I agree that requiring header offset to be at zero would simplify the use, but I don't like the limitation it does add.
- Another solution (I had this in mind for later enhancement) is to use the offset in `Message` itself to indicate start of DNS header (this way we keep it flexible while simplifying the APIs).

Reply:
> Another solution (I had this in mind for later enhancement) is to use the offset in Message itself to indicate start of DNS header (this way we keep it flexible while simplifying the APIs).

Having the message object carry its own offset value was primarily intended to serve use cases like this.